maturity-level-two/src/Controllers/v1/CustomizationController.cs

```diff
@@ -0,0 +1,186 @@
+        {
+            var CustomizationObj = await _coditoRepository.GetCustomizationAsync(id);
+            if (CustomizationObj == null)
+            {
+                return NotFound(new ProblemDetailsError(StatusCodes.Status404NotFound));
+            }
+            await _coditoRepository.DeleteCustomizationAsync(id);
+            return NoContent();
+        }
+
+        /// <summary>
+        /// Update customization
+        /// </summary>
+        /// <param name="id">customization identifier</param>
+        /// <param name="customizationDto">Update for the customization object</param>
+        /// <remarks>Update the customization profile (incremental update)</remarks>
+        /// <returns>Acknowledge the object has been updated</returns>
+        [HttpPatch("{id}", Name = Constants.RouteNames.v1.UpdateCustomizationIncremental)]
+        [SwaggerResponse((int)HttpStatusCode.NoContent, "The data has been updated")]
+        [SwaggerResponse((int)HttpStatusCode.NotFound, "Customization not found")]
+        [SwaggerResponse((int)HttpStatusCode.InternalServerError, "API is not available")]
+        public async Task<IActionResult> UpdateIncremental(int id, [FromBody] CustomizationDto customizationDto)
+        {
+            var CustomizationObj = await _coditoRepository.GetCustomizationAsync(id);
+            if (CustomizationObj == null)
+            {
+                return NotFound(new ProblemDetailsError(StatusCodes.Status404NotFound));
+            }
+
+            var customizationUpdated = Mapper.Map<Customization>(customizationDto);
+            customizationUpdated.Id = id;
+
+            await _coditoRepository.ApplyPatchAsync<Customization, CustomizationDto>(customizationUpdated, customizationDto);
+            return NoContent();
+
+        }
+
+        /// <summary>
+        /// Sell a car customization
+        /// </summary>
+        /// <param name="id">Customization identifier</param>
+        /// <returns>Acknowledge that the c</returns>
+        [HttpPost("{id}/sale", Name = Constants.RouteNames.v1.SellCustomization)]
+        [SwaggerResponse((int)HttpStatusCode.Accepted, "Sale accepted. No response body")]
+        [SwaggerResponse((int)HttpStatusCode.BadRequest, "Out of stock")]
+        [SwaggerResponse((int)HttpStatusCode.NotFound, "Car customization not found")]
+        [SwaggerResponse((int)HttpStatusCode.InternalServerError, "API is not available")]
+        public async Task<IActionResult> VoteAsBestPlayer(int id)
+        {
+            var customization = await _coditoRepository.GetCustomizationAsync(id);
+            if (customization == null)
+            {
+                return NotFound(new ProblemDetailsError(StatusCodes.Status404NotFound));
+            }
+            if (customization.InventoryLevel <= 0)
+            {
+                return BadRequest();
+            }
+
+            await _coditoRepository.ApplyCustomizationSaleAsync(customization);
```

Comment:
Actually, all these operations on the database should be done inside a (exclusive) transaction, since a race-condition is introduced here.

It is possible that multiple sales requests for the same item are coming in at the same time.  User A can retrieve the customization to check the inventory level, and that might be fine but by the time the `ApplyCustomizationSaleAsync` method is executed, the inventory level might already be changed to 0 by other incoming requests.

Reply:
issue created